### PR TITLE
RavenDB-11747 => make sure ByteStringMemoryCache properly uses NativeMemoryCleaner

### PR DESCRIPTION
**Pull request by myarichuk**
Because the ByteStringMemoryCache was a struct, it passed the wrong reference to RegisterLowMemoryHandler(), thus not executing the low memory event when needed.

### CODE REVIEW

File: src/Raven.Server/Program.cs [hunk 1]
```diff
@@ -317,6 +319,82 @@ private static bool RunInteractive(RavenServer server)
...
             Console.WriteLine("Showing stats, press any key to close...");
+            ContinuouslyWriteServerStats(server);
+
+            Console.ReadKey(true);
+            Console.WriteLine();
+            Console.WriteLine($"Stats halted.");
+        }
+
+        private static readonly Random Random = new Random();
+        public static  (
```
Comment (ayende) [inline src/Raven.Server/Program.cs:330]: This is ugly, let's introduce an actual struct/class for this

File: src/Sparrow/ByteString.cs [hunk 2]
```diff
@@ -467,7 +467,7 @@ public interface IByteStringAllocator
     /// <summary>
     /// This class implements a direct allocator, mostly used for testing.   
     /// </summary>
-    public struct ByteStringDirectAllocator : IByteStringAllocator
+    public class ByteStringDirectAllocator : IByteStringAllocator
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:470]: This is a _big_ change!
@redknightlois Can you also comment on this?

File: src/Sparrow/ByteString.cs [hunk 3]
```diff
@@ -679,7 +681,7 @@ public int SizeLeft
                 get { return (int)(End - Current); }
             }
         }
-
+        private static readonly TAllocator Allocator = new TAllocator();
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:684]: I'm pretty sure that this isn't thread safe.

File: src/Sparrow/ByteString.cs [hunk 4]
```diff
@@ -494,8 +494,12 @@ static ByteStringMemoryCache()
             Cleaner = new NativeMemoryCleaner<SegmentStack, UnmanagedGlobalSegment>(SegmentsPool, LowMemoryFlag, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1));
 
             ThreadLocalCleanup.ReleaseThreadLocalState += CleanForCurrentThread;
+        }
 
-            LowMemoryNotification.Instance.RegisterLowMemoryHandler(ByteStringContext.Allocator);
+        public ByteStringMemoryCache()
+        {
+            //each instance of this should handle low memory...
+            LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:502]: Note that the state here is all completely static. Better to create a dedicated class that would do the cleanup than change these to structs, I think

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/ByteString.cs (hunk 2 above)
Comment (redknightlois) [reply src/Sparrow/ByteString.cs:470]: This will kill the performance of the allocator.

**Comment by redknightlois** [comment]
You need to register the context itself to propagate the low memory signal. That is how it is being done on the generalized allocators.

**Pull request closed without merging**